bls12_381/src/g2.rs

```diff
@@ -885,6 +923,385 @@ impl G2Projective {
+    fn from(val: G2Subgroup) -> G2Projective {
+        val.0
+    }
+}
+
+impl fmt::Display for G2Subgroup {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl Sum for G2Subgroup {
+    fn sum<I: Iterator<Item = Self>>(iter: I) -> Self {
+        iter.fold(Self::identity(), Add::add)
+    }
+}
+
+impl<'r> Sum<&'r G2Subgroup> for G2Subgroup {
+    fn sum<I: Iterator<Item = &'r Self>>(iter: I) -> Self {
+        iter.fold(Self::identity(), Add::add)
+    }
+}
+
+impl Neg for G2Subgroup {
+    type Output = G2Subgroup;
+
+    #[inline]
+    fn neg(self) -> G2Subgroup {
+        G2Subgroup(-self.0)
+    }
+}
+
+impl<'a, 'b> Add<&'b G2Subgroup> for &'a G2Projective {
+    type Output = G2Projective;
+
+    #[inline]
+    fn add(self, other: &'b G2Subgroup) -> G2Projective {
+        self + &other.0
+    }
+}
+
+impl<'a, 'b> Sub<&'b G2Subgroup> for &'a G2Projective {
+    type Output = G2Projective;
+
+    #[inline]
+    fn sub(self, other: &'b G2Subgroup) -> G2Projective {
+        self - &other.0
+    }
+}
+
+impl_binops_additive!(G2Projective, G2Subgroup);
+
+impl<'a, 'b> Add<&'b G2Subgroup> for &'a G2Subgroup {
+    type Output = G2Subgroup;
+
+    #[inline]
+    fn add(self, other: &'b G2Subgroup) -> G2Subgroup {
+        G2Subgroup(self.0 + &other.0)
+    }
+}
+
+impl<'a, 'b> Sub<&'b G2Subgroup> for &'a G2Subgroup {
+    type Output = G2Subgroup;
+
+    #[inline]
+    fn sub(self, other: &'b G2Subgroup) -> G2Subgroup {
+        G2Subgroup(self.0 - &other.0)
+    }
+}
+
+impl_binops_additive!(G2Subgroup, G2Subgroup);
+
+impl<'a, 'b> Mul<&'b Scalar> for &'a G2Subgroup {
+    type Output = G2Subgroup;
+
+    fn mul(self, other: &'b Scalar) -> G2Subgroup {
+        G2Subgroup(self.0.multiply(&other.to_bytes()))
+    }
+}
+
+impl_binops_multiplicative!(G2Subgroup, Scalar);
+
+pub struct G2Compressed([u8; 96]);
+
+impl fmt::Debug for G2Compressed {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.0[..].fmt(f)
+    }
+}
+
+impl Default for G2Compressed {
+    fn default() -> Self {
+        G2Compressed([0; 96])
+    }
+}
+
+impl AsRef<[u8]> for G2Compressed {
+    fn as_ref(&self) -> &[u8] {
+        &self.0
+    }
+}
+
+impl AsMut<[u8]> for G2Compressed {
+    fn as_mut(&mut self) -> &mut [u8] {
+        &mut self.0
+    }
+}
+
+pub struct G2Uncompressed([u8; 192]);
+
+impl fmt::Debug for G2Uncompressed {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.0[..].fmt(f)
+    }
+}
+
+impl Default for G2Uncompressed {
+    fn default() -> Self {
+        G2Uncompressed([0; 192])
+    }
+}
+
+impl AsRef<[u8]> for G2Uncompressed {
+    fn as_ref(&self) -> &[u8] {
+        &self.0
+    }
+}
+
+impl AsMut<[u8]> for G2Uncompressed {
+    fn as_mut(&mut self) -> &mut [u8] {
+        &mut self.0
+    }
+}
+
+impl Group for G2Projective {
+    type Scalar = Scalar;
+
+    fn random<R: RngCore + ?Sized>(rng: &mut R) -> Self {
```

Comment:
Let's document in the `random` API definition in the `Group` trait that this is not guaranteed to be deterministic.